File: 2016/Prototypes/CollisionTest/src/com/kennedyrobotics/robot2016/subsystems/SensorSubsystem.java
package com.kennedyrobotics.robot2016.subsystems;

import com.kauailabs.navx.frc.AHRS;
import com.kennedyrobotics.robot2016.RobotMap;
import com.kennedyrobotics.robot2016.commands.SensorUpdateCommand;

import edu.wpi.first.wpilibj.DigitalInput;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.SerialPort;
import edu.wpi.first.wpilibj.command.Subsystem;
import edu.wpi.first.wpilibj.livewindow.LiveWindow;

/**
 *
 */
public class SensorSubsystem extends Subsystem {
    
	public static final double DRIVE_INCH_PER_COUNT = 0.1570796327;
	
    // Put methods for controlling this subsystem
    // here. Call these from Commands.

	private final DigitalInput wallDetector = new DigitalInput(RobotMap.CHASSIS_WALL_DEDECTOR);
	
	private AHRS navX;
	private boolean collision = false;
		
	public SensorSubsystem() {
		
		try {
			navX = new AHRS(SerialPort.Port.kUSB);
		} catch (Exception e) {
			DriverStation.reportError("Error instantiating navX MXP:  " + e.getMessage(), true);
			navX = null;
		}
		if(navX != null) {
			LiveWindow.addSensor("IMU", "Gyro", navX);
		}
	}
	
    public void initDefaultCommand() {
    	setDefaultCommand(new SensorUpdateCommand());
    }
    
    public AHRS getNavX() {
    	return navX;
    }
    
	public double getYaw() {
		if(navX == null) {
			DriverStation.reportError("NavX is null", false);
			return 0.0;
		} else {
			return navX.getYaw();

		}
	}
	
	public void setCollision(boolean collision){
		this.collision = collision;
	}
	
	public boolean getCollision(){
		return collision;
	}
	
	public boolean atWall() {
		return !wallDetector.get();
	}
}

File: 2017/2017Robot/src/Subsystems/SensorSubsystem.cpp
#include "SensorSubsystem.h"
#include "../RobotMap.h"
#include "../RobotConstants.h"

#include <DigitalInput.h>
#include <math.h>
#include "Chassis.h"

#include "Commands/SensorUpdateCommand.h"
#include <AHRS.h>

SensorSubsystem::SensorSubsystem() : Subsystem("SensorSubsystem") {

	pdp = new PowerDistributionPanel();

	climberIndex = new DigitalInput(CLIMBER_INDEX);

	chassisLeftEncoder = new Encoder(CHASSIS_LEFT_ENCODER_A, CHASSIS_LEFT_ENCODER_B, Encoder::k4X);
	chassisLeftEncoder->SetPIDSourceType(PIDSourceType::kDisplacement);
	chassisLeftEncoder->SetDistancePerPulse(kCHASSIS_DISTANCE_PER_PULSE);
	chassisLeftEncoder->SetReverseDirection(true);
	chassisLeftEncoder->Reset();

	chassisRightEncoder = new Encoder(CHASSIS_RIGHT_ENCODER_A, CHASSIS_RIGHT_ENCODER_B, Encoder::k4X);
	chassisRightEncoder->SetPIDSourceType(PIDSourceType::kDisplacement);
	chassisRightEncoder->SetDistancePerPulse(kCHASSIS_DISTANCE_PER_PULSE);
	chassisRightEncoder->SetReverseDirection(false);
	chassisRightEncoder->Reset();

	gyro = new AnalogGyro(1);

#ifdef ENABLE_IMU
	try {
		imu = new AHRS(SerialPort::Port::kUSB);
		DriverStation::ReportWarning("The MXP is initialized, we think.");
	} catch(std::exception &ex) {
		std::string err_string = "Error instantiating navX MXP:  ";
		err_string += ex.what();
		DriverStation::ReportError(err_string.c_str());
	}

#else
	imu = NULL;
#endif
}

void SensorSubsystem::InitDefaultCommand() {
	// Set the default command for a subsystem here.
	// SetDefaultCommand(new MySpecialCommand());
	SetDefaultCommand(new SensorUpdateCommand());
}

// Put methods for controlling this subsystem
// here. Call these from Commands.
AHRS* SensorSubsystem::GetIMU() {
	return imu;
}

PowerDistributionPanel* SensorSubsystem::GetPDP() {
	return this->pdp;
}

Encoder* SensorSubsystem::GetChassisLeftEncoder() {
	return chassisLeftEncoder;
}

Encoder* SensorSubsystem::GetChassisRightEncoder() {
	return chassisRightEncoder;
}

double SensorSubsystem::GetChassisLeftCurrent() {
	return (pdp->GetCurrent(PDP_CHASSIS_LEFT_FRONT)+ pdp->GetCurrent(PDP_CHASSIS_LEFT_FRONT))/2.0;
}

double SensorSubsystem::GetChassisRightCurrent() {
	return (pdp->GetCurrent(PDP_CHASSIS_RIGHT_FRONT)+ pdp->GetCurrent(PDP_CHASSIS_RIGHT_FRONT))/2.0;
}

bool SensorSubsystem::GetClimberIndex() {
	return !climberIndex->Get();
}

double SensorSubsystem::GetYaw() {
#ifdef SELECT_NAVX
	return imu->GetYaw();
#else
	return gyro->GetAngle();
#endif
}

double SensorSubsystem::GetYawRate() {
#ifdef SELECT_NAVX
	return imu->GetRate();
#else
	return gyro->GetRate();
#endif
}

void SensorSubsystem::ZeroYaw() {
#ifdef SELECT_NAVX
	return imu->ZeroYaw();
#else
	return gyro->Reset();
#endif
}

PIDSource* SensorSubsystem::GetYawSource() {
#ifdef SELECT_NAVX
	return imu;
#else
	return gyro;
#endif

}

void SensorSubsystem::ResetPhysicalPose() {
	chassisLeftEncoder->Reset();
	chassisRightEncoder->Reset();
	this->ZeroYaw();
}

Pose SensorSubsystem::GetPhysicalPose() {
	return Pose(chassisLeftEncoder->GetDistance(), chassisRightEncoder->GetDistance(),
			chassisLeftEncoder->GetRate(), chassisRightEncoder->GetRate(),
			this->GetYaw(), this->GetYawRate());
}
